### ISSUE TITLE
[C++][Parquet] Forced UTF8 encoding of BYTE_ARRAY on stream::read/write

### ISSUE THREAD

**Issue by asfimport** [apache/arrow#42971]
```java

StreamReader& StreamReader::operator>>(optional<std::string>& v) {
 CheckColumn(Type::BYTE_ARRAY, ConvertedType::UTF8);
 ByteArray ba;
```
 
```java

StreamWriter& StreamWriter::WriteVariableLength(const char* data_ptr,
 std::size_t data_len) {
 CheckColumn(Type::BYTE_ARRAY, ConvertedType::UTF8);
```
 

Though the C++ Parquet::Schema::Node allows physical type of BYTE_ARRAY with ConvertedType=NONE, the stream reader/writer classes throw when ConvertedType != UTF8.

std::string is, unfortunately, the canonical byte buffer class in C++.

A simple approach might be to create >>parquet::ByteArray.. with columnCheck(BYTE_ARRAY, NONE), and let the user take it from there.  that would use the existing methods that >>std::string uses.. just an idea.

I am new to this forum, and have assigned MAJOR to this bug, but gladly defer to those who have a better grasp of classification.

**Reporter**: [ian](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=protalis)

<sub>**Note**: *This issue was originally created as [PARQUET-1958](https://issues.apache.org/jira/browse/PARQUET-1958). Please see the [migration documentation](https://issues.apache.org/jira/browse/PARQUET-2502) for further details.*</sub>

**Comment by asfimport** [comment]
[Micah Kornfield](https://issues.apache.org/jira/browse/PARQUET-1958?#comment-17270548) / @emkornfield:
I actually am not sure that the check is needed at all.  It seems any byte array should be readable at this point.  My memory of this code is a little bit hazy, but do you think there is harm in removing the check?

**Comment by asfimport** [comment]
[ian](https://issues.apache.org/jira/browse/PARQUET-1958?#comment-17289541):
hi micah.  no.. it should be removed to align with schema.node checks and to support unencoded byte arrays.  thank you for following up.  best, ian.Sent from my Verizon, Samsung Galaxy smartphone
-------- Original message --------From: "Micah Kornfield (Jira)" <jira@apache.org> Date: 1/22/21  22:27  (GMT-06:00) To: com@ianlavie.com Subject: [jira] [Commented] (PARQUET-1958) Forced UTF8 encoding of
  BYTE_ARRAY on stream::read/write     [ https://issues.apache.org/jira/browse/PARQUET-1958?page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel&focusedCommentId=17270548#comment-17270548 ] Micah Kornfield commented on PARQUET-1958:-----------------------------------------~~I actually am not sure that the check is needed at all.  It seems any byte array should be readable at this point.  My memory of this code is a little bit hazy, but do you think there is harm in removing the check?> Forced UTF8 encoding of BYTE_ARRAY on stream::read/write> -~~------------------------------------------------------>>                 Key: PARQUET-1958>                 URL: https://issues.apache.org/jira/browse/PARQUET-1958>             Project: Parquet>          Issue Type: Bug>          Components: parquet-cpp>    Affects Versions: cpp-1.5.0>            Reporter: ian>            Priority: Major>> ```java
> StreamReader& StreamReader::operator>>(optional<std::string>& v) {>  CheckColumn(Type::BYTE_ARRAY, ConvertedType::UTF8);>  ByteArray ba;
```>  > ```java
> StreamWriter& StreamWriter::WriteVariableLength(const char* data_ptr,>  std::size_t data_len) {>  CheckColumn(Type::BYTE_ARRAY, ConvertedType::UTF8);
```>  > Though the C++ Parquet::Schema::Node allows physical type of BYTE_ARRAY with ConvertedType=NONE, the stream reader/writer classes throw when ConvertedType != UTF8.> std::string is, unfortunately, the canonical byte buffer class in C++.> A simple approach might be to create >>parquet::ByteArray.. with columnCheck(BYTE_ARRAY, NONE), and let the user take it from there.  that would use the existing methods that >>std::string uses.. just an idea.> I am new to this forum, and have assigned MAJOR to this bug, but gladly defer to those who have a better grasp of classification.--This message was sent by Atlassian Jira(v8.3.4#803005)

**Comment by asfimport** [comment]
[Antoine Pitrou](https://issues.apache.org/jira/browse/PARQUET-1958?#comment-17289822) / @pitrou:
Indeed the UTF8 check doesn't seem necessary.

**Comment by pitrou** [comment]
Issue resolved by pull request 44739
https://github.com/apache/arrow/pull/44739

**Comment by pitrou** [comment]
@pulkomandy Can you post a comment on this issue so that we can assign it to you?

**Comment by pitrou** [comment]
@pulkomandy By the way, you may be fine with this, but just be aware that the stream reader and writer are not really high-performance APIs for accessing Parquet.

**Comment by pulkomandy** [comment]
It is working well enough for our usage for a few years already. I think we don't really have other choices for what we're doing (live data collection in an embedded Linux system). The non-streaming APIs would require us to first store the data in some other format, if I remember correctly. The extra IO cost for that (writing in another format to mass storage and reading it back to convert in one go) would not be so great.

**Comment by pitrou** [comment]
Yes, I suppose you could use a `TypedColumnWriter` and write small batches of a hundred values or so, but perhaps that wouldn't be much faster anyway.

(it will also depend heavily on the options used for writing: encoding, compression, etc.)
